File: packages/gbnf-experiment/data/20260909T123239Z_5982b634/ported_implementation/src/grammar-graph/graph-pointer.ts
import type { GraphNode } from './graph-node.js';
import { printGraphPointer } from './print.js';
import { isRuleChar, isRuleCharExclude, isRuleEnd, isRuleRef } from './type-guards.js';
import type { PrintOpts, ResolvedGraphPointer, UnresolvedRule } from './types.js';

export type GraphPointerKey = string;

export class GraphPointer<T extends UnresolvedRule = UnresolvedRule> {
  node: GraphNode<T>;
  parent?: GraphPointer;
  id: GraphPointerKey;
  #valid?: boolean;

  constructor(node: GraphNode<T>, parent?: GraphPointer) {
    if (!node) {
      throw new Error('Node is undefined');
    }
    this.node = node;
    this.parent = parent;
    this.id = parent ? `${parent.id}-${node.id}` : node.id;
  }

  get rule(): T {
    return this.node.rule;
  }

  get valid(): boolean | undefined {
    return this.#valid;
  }

  set valid(valid: boolean | undefined) {
    this.#valid = valid;
  }

  print(opts: PrintOpts): string {
    return printGraphPointer(this)(opts);
  }

  /**
   * Resolve this pointer into pointers that sit on concrete (non-reference) rules.
   *
   * @param resolved whether the pointer has already been resolved
   */
  *resolve(resolved = false): Generator<ResolvedGraphPointer> {
    const rule = this.rule;
    if (isRuleRef(rule)) {
      if (resolved) {
        if (!this.node.next) {
          throw new Error(`No next node: ${JSON.stringify(this.node.meta)}`);
        }
        yield* new GraphPointer(this.node.next, this.parent).resolve();
      } else {
        for (const node of rule.nodes) {
          yield* new GraphPointer(node, this).resolve();
        }
      }
    } else if (isRuleEnd(rule)) {
      if (!this.parent) {
        yield this as unknown as ResolvedGraphPointer;
      } else {
        yield* this.parent.resolve(true);
      }
    } else if (isRuleChar(rule) || isRuleCharExclude(rule)) {
      yield this as unknown as ResolvedGraphPointer;
    } else {
      throw new Error(`Unknown rule: ${JSON.stringify(rule)}`);
    }
  }

  /**
   * Fetch the pointers that follow this one, if this pointer is still valid.
   */
  *fetchNext(): Generator<GraphPointer> {
    // If this pointer is invalid, then we don't return any new pointers
    if (!this.valid) {
      return;
    }

    // If this pointer is an end node, we return the parent's next node.
    // If no parent exists, we return nothing, since it's the end of the line.
    if (isRuleEnd(this.node.rule)) {
      if (this.parent) {
        yield* this.parent.fetchNext();
      }
    } else {
      if (!this.node.next) {
        throw new Error(`No next node: ${JSON.stringify(this.node.meta)}`);
      }
      yield* new GraphPointer(this.node.next, this.parent).resolve();
    }
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T123239Z_5982b634/20260910T070936Z_fb630842/ported_implementation/gbnf/grammar_graph/graph_pointer.py
from __future__ import annotations

from typing import Iterator, Optional

from .print import print_graph_pointer
from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref
from .types import PrintOpts


class GraphPointer:
    def __init__(self, node, parent: Optional['GraphPointer'] = None):
        if node is None:
            raise ValueError('Node is undefined')
        self.node = node
        self.parent = parent
        self.id: str = f'{parent.id}-{node.id}' if parent is not None else node.id
        self.valid: Optional[bool] = None

    @property
    def rule(self):
        return self.node.rule

    def print(self, opts: PrintOpts) -> str:
        return print_graph_pointer(self)(opts)

    def resolve(self, resolved: bool = False) -> Iterator['GraphPointer']:
        """Resolve this pointer into pointers that sit on concrete (non-reference) rules.

        :param resolved: whether the pointer has already been resolved
        """
        rule = self.rule
        if is_rule_ref(rule):
            if resolved:
                if self.node.next is None:
                    raise ValueError(f'No next node: {self.node.meta!r}')
                yield from GraphPointer(self.node.next, self.parent).resolve()
            else:
                for node in rule.nodes:
                    yield from GraphPointer(node, self).resolve()
        elif is_rule_end(rule):
            if self.parent is None:
                yield self
            else:
                yield from self.parent.resolve(True)
        elif is_rule_char(rule) or is_rule_char_exclude(rule):
            yield self
        else:
            raise ValueError(f'Unknown rule: {rule!r}')

    def fetch_next(self) -> Iterator['GraphPointer']:
        """Fetch the pointers that follow this one, if this pointer is still valid."""
        # If this pointer is invalid, then we don't return any new pointers
        if not self.valid:
            return

        # If this pointer is an end node, we return the parent's next node.
        # If no parent exists, we return nothing, since it's the end of the line.
        if is_rule_end(self.node.rule):
            if self.parent is not None:
                yield from self.parent.fetch_next()
        else:
            if self.node.next is None:
                raise ValueError(f'No next node: {self.node.meta!r}')
            yield from GraphPointer(self.node.next, self.parent).resolve()

    def __repr__(self) -> str:
        return f'GraphPointer({self.id})'
